### PR TITLE
feat(carousel): L3-3145 carousel component in seldon

### PR DESCRIPTION
**Pull request by sgrund14**
**Jira ticket**

[L3-3145](https://phillipsauctions.atlassian.net/browse/L3-3145)

**Screenshots**
| Before | After |
| ------ | ----- |
| Paste Before Image | Paste After Image |

**Mobile**
https://github.com/user-attachments/assets/cac28f88-5724-4e58-9a3e-f69a56d1bba9

**Desktop**
https://github.com/user-attachments/assets/f855d723-8351-4ddb-bb4e-cabaa387082a

**Summary**
Composable Carousel component suite in Seldon

Still awaiting: overflow behavior from design, loading state, null state

Add high-level overview of the changes that were made and explanation of WHY they were made

**Change List (describe the changes made to the files)**

- Carousel, CarouselContent, CarouselItem, and CarouselDots components

**Acceptance Test (how to verify the PR)**

- Add step by step instructions on how to verify the change

**Regression Test**

- (Optional) Add verification steps to make sure this PR doesn't break old functionality

**Evidence of testing**

- Post logs, screenshots, etc

**Things to look for during review**

- [ ] PR title should correctly describe the most significant type of commit. I.e. `feat(scope): ...` if a `minor` release should be triggered.
- [ ] All commit messages follow convention and are appropriate for the changes
- [ ] All references to `phillips` class prefix are using the prefix variable
- [ ] All major areas have a `data-testid` attribute.
- [ ] Document all props with jsdoc comments
- [ ] All strings should be translatable.
- [ ] Unit tests should be written and should have a coverage of 90% or higher in all areas.

New Components

- [ ] Are there any [accessibility considerations](https://www.w3.org/WAI/ARIA/apg/patterns/) that need to be taken into account and tested?
- [ ] Default story called "Playground" should be created for all new components
- [ ] Create a jsdoc comment that has an Overview section and a link to the Figma design for the component
- [ ] Export the component and its typescript type from the `index.ts` file
- [ ] Import the component scss file into the `componentStyles.scss` file.

[L3-3145]: https://phillipsauctions.atlassian.net/browse/L3-3145?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by constantinehuzenko** [comment]
dots are weirdly moving on click and on slide 

https://github.com/user-attachments/assets/f21181c0-9b67-4120-a563-7718b6bd2d9b

**Review by constantinehuzenko — CHANGES_REQUESTED** [review]
other then dots problem I don't see any code problems, I would maybe crate another custom hook to cover all these onSelect and useEffects logic, so the components know less. But we don't have standard for that so I guess we good to go as soon as dots fixed

**Comment by scottdickerson** [comment]
is there a way to standardize the `gap` with a prop tied to "SpacingTokens" between the images in the carousel when swiping left and right?  They're currently butted right up against each other

File: src/components/Carousel/Carousel.stories.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+const meta = {
+  title: 'Components/Carousel',
+  component: Carousel,
+} satisfies Meta<typeof Carousel>;
+
+export default meta;
+export const Playground = (props: CarouselProps) => (
+  <Carousel {...props}>
+    <CarouselContent>
+      {Array.from({ length: 9 }).map((_, index) => (
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.stories.tsx:17]: I always forget this, nice

File: src/components/Carousel/Carousel.test.tsx [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import CarouselItem from './CarouselItem';
+import CarouselDots from './CarouselDots';
+import { runCommonTests } from '../../utils/testUtils';
+
+describe('Carousel', () => {
+  runCommonTests(Carousel, 'Carousel');
+
+  beforeEach(() => {
+    class MockIntersectionObserver {
+      observe = () => {}; // eslint-disable-line @typescript-eslint/no-empty-function
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.test.tsx:13]: also nice, perhaps we should move this stuff to the global test setup file so all testcases get it https://github.com/PhillipsAuctionHouse/seldon/blob/c4a909a1ac54aeffe93988c5adb83cd9a08a71cc/config/vitest/setupTest.ts#L16

File: src/components/Carousel/Carousel.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      if (!api) {
+        return;
+      }
+
+      setCanScrollPrev(api.canScrollPrev());
+      setCanScrollNext(api.canScrollNext());
+    }, []);
+
+    const scrollPrev = useCallback(() => {
+      api?.scrollPrev();
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.tsx:72]: the extra `useCallback` might not be needed here as we shouldn't have too many children, we would have to `React.memo` the children to make this work anyway

File: src/components/Carousel/Carousel.tsx [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+    const handleKeyDown = useCallback(
+      (event: KeyboardEvent<HTMLDivElement>) => {
+        if (event.key === 'ArrowLeft') {
+          event.preventDefault();
+          scrollPrev();
+        } else if (event.key === 'ArrowRight') {
+          event.preventDefault();
+          scrollNext();
+        }
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.tsx:87]: will you need to trigger the `onClick` on "Enter" ?

File: src/components/Carousel/Carousel.tsx [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+    useEffect(() => {
+      if (!api) {
+        return;
+      }
+      if (onSlideChange) {
+        api.on('select', (embla) => {
+          onSlideChange(embla.selectedScrollSnap());
+        });
+      }
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.tsx:100]: need to cleanup the `api.on` with `api.off` on return

File: src/components/Carousel/Carousel.tsx [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    }, [api, onSlideChange]);
+
+    useEffect(() => {
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.tsx:103]: I'm not sure why you need this one?

File: src/components/Carousel/Carousel.tsx [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    const { className: baseClassName, ...commonProps } = getCommonProps(props, 'Carousel');
+
+    const [carouselRef, api] = useEmblaCarousel({
+      loop,
+      startIndex,
+    });
+    const [canScrollPrev, setCanScrollPrev] = useState(false);
+    const [canScrollNext, setCanScrollNext] = useState(false);
+
+    const onSelect = useCallback((api: CarouselApi) => {
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.tsx:62]: if you can remove the `useState` above you can remove this `onSelect`

File: src/components/Carousel/Carousel.tsx [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      if (!api) {
+        return;
+      }
+
+      onSelect(api);
+      api.on('reInit', onSelect);
+      api.on('select', onSelect);
+
+      return () => {
+        api?.off('select', onSelect);
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.tsx:113]: need `api.off('reInit'` on return?

File: src/components/Carousel/Carousel.tsx [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+          canScrollPrev,
+          canScrollNext,
+        }}
+      >
+        <div
+          ref={ref}
+          onKeyDownCapture={handleKeyDown}
+          className={classnames(baseClassName, className)}
+          role="region"
+          aria-roledescription="carousel"
```
Comment (scottdickerson) [inline src/components/Carousel/Carousel.tsx:133]: nice!

File: src/components/Carousel/CarouselDots.tsx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+/**
+ * ## Overview
+ *
+ * Carousel dots component that displays pagination for the carousel.
+ * It supports a maximum number of visible dots
+ * The dots animate left to right as the user navigates through the carousel.
+ *
+ */
+const CarouselDots = forwardRef<HTMLDivElement, CarouselDotsProps>(
+  ({ className, maxDots = 7, position = 'inline', ...props }, ref) => {
```
Comment (scottdickerson) [inline src/components/Carousel/CarouselDots.tsx:27]: if you have 20 images, and maxDots is set to 7, when you get to the edge of the dots, what happens?  how can you go from image 7 to image 8 with dots?

File: src/components/Carousel/Carousel.stories.tsx (hunk 1 above)
Comment (sgrund14) [reply src/components/Carousel/Carousel.stories.tsx:17]: convenient

File: src/components/Carousel/Carousel.test.tsx (hunk 2 above)
Comment (sgrund14) [reply src/components/Carousel/Carousel.test.tsx:13]: down 👍

File: src/components/Carousel/Carousel.tsx [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+ */
+const Carousel = forwardRef<HTMLDivElement, CarouselProps>(
+  ({ loop = false, startIndex = 0, onSlideChange, className, children, ...props }, ref) => {
+    const { className: baseClassName, ...commonProps } = getCommonProps(props, 'Carousel');
+
+    const [carouselRef, api] = useEmblaCarousel({
+      loop,
+      startIndex,
+    });
+    const [canScrollPrev, setCanScrollPrev] = useState(false);
```
[the remark this answers is not in the document]
Comment (sgrund14) [reply src/components/Carousel/Carousel.tsx:59]: I think there isn't a re-render cycle whenever you navigate, so you have to manually do it in the onSelect callback

[the remark this answers is not in the document]
Comment (sgrund14) [reply src/components/Carousel/Carousel.tsx:59]: actually JK i'm wrong

File: src/components/Carousel/Carousel.tsx (hunk 4 above)
Comment (sgrund14) [reply src/components/Carousel/Carousel.tsx:87]: think that could be handled by the pagination / the slides themselves?

File: src/components/Carousel/utils.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import { useContext } from 'react';
+import { CarouselContext } from './Carousel';
+
+export function useCarousel() {
```
Comment (scottdickerson) [inline src/components/Carousel/utils.ts:4]: this is a great pattern

Comment (scottdickerson) [reply src/components/Carousel/utils.ts:4]: to make a separate named hook

File: src/components/Carousel/types.ts [hunk 13]
```diff
@@ -0,0 +1 @@
+// Add any additional typescript types here
```
Comment (scottdickerson) [inline src/components/Carousel/types.ts:1]: can remove this file

File: src/components/Carousel/_carousel.scss [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+    flex: 0 0 100%;
+    min-width: 0;
+
+    @each $spacing-size in $spacing-sizes {
+      &--gap-#{$spacing-size} {
+        padding-left: var(--spacing-#{$spacing-size});
+      }
+    }
+
+    // &:not(:first-child) {
```
Comment (scottdickerson) [inline src/components/Carousel/_carousel.scss:32]: can remove?

File: src/components/Carousel/_carousel.scss [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+  position: relative;
+
+  &-content {
+    overflow: hidden;
+  }
+
+  &-content-inner {
+    display: flex;
+
+    // margin-left: -1rem; // Equivalent to -ml-4
```
Comment (scottdickerson) [inline src/components/Carousel/_carousel.scss:13]: can remove?

File: src/components/Carousel/Carousel.tsx (hunk 4 above)
Comment (scottdickerson) [reply src/components/Carousel/Carousel.tsx:73]: for the zoom feature?

**Review by scottdickerson — APPROVED** [review]
LGTM

File: src/components/Carousel/Carousel.tsx (hunk 4 above)
Comment (sgrund14) [reply src/components/Carousel/Carousel.tsx:73]: think that'd get handled by the consumer, I'll open a PR with my zoom implementation

File: src/components/Carousel/utils.ts (hunk 12 above)
Comment (sgrund14) [reply src/components/Carousel/utils.ts:4]: yea it's nice to pass around

**Review by constantinehuzenko — APPROVED** [review]
LGTM

**Pull request merged**